components/service/glean/src/main/java/mozilla/components/service/glean/storages/CustomDistributionsStorageEngine.kt

```diff
@@ -0,0 +1,407 @@
+    val count: Long
+        get() = values.map { it.value }.sum()
+
+    // This is a helper property to build the correct identifier for the metric and allow for
+    // blank categories
+    internal val identifier: String = if (category.isEmpty()) { name } else { "$category.$name" }
+
+    // This is a list of limits for the buckets.  Instantiated lazily to ensure that the range and
+    // bucket counts are set first.
+    internal val buckets: List<Long> by lazy { getBuckets() }
+
+    /**
+     * Finds the correct bucket, using a binary search to locate the index of the
+     * bucket where the sample is bigger than or equal to the bucket limit.
+     *
+     * @param sample Long value representing the sample that is being accumulated
+     */
+    internal fun findBucket(sample: Long): Long {
+        var under = 0
+        var over = bucketCount
+        var mid: Int
+
+        do {
+            mid = under + (over - under) / 2
+            if (mid == under) {
+                break
+            }
+
+            if (buckets[mid] <= sample) {
+                under = mid
+            } else {
+                over = mid
+            }
+        } while (true)
+
+        return buckets[mid]
+    }
+
+    /**
+     * Accumulates a sample to the correct bucket.
+     * If a value doesn't exist for this bucket yet, one is created.
+     *
+     * @param sample Long value representing the sample that is being accumulated
+     */
+    internal fun accumulate(sample: Long) {
+        val limit = findBucket(sample)
+        values[limit] = (values[limit] ?: 0) + 1
+        sum += sample
+    }
+
+    /**
+     * Helper function to build the [CustomDistributionData] into a JSONObject for serialization
+     * purposes.
+     */
+    internal fun toJsonObject(): JSONObject {
+        return JSONObject(mapOf(
+            "category" to category,
+            "name" to name,
+            "bucket_count" to bucketCount,
+            "range" to JSONArray(arrayOf(rangeMin, rangeMax)),
+            "histogram_type" to histogramType.toString().toLowerCase(),
+            "values" to values.mapKeys { "${it.key}" },
+            "sum" to sum
+        ))
+    }
+
+    /**
+     * Helper function to build the [CustomDistributionData] into a JSONObject for sending in the
+     * ping payload. Compared to [toJsonObject] which is designed for lossless roundtripping:
+     *
+     *   - this does not include the bucketing parameters
+     *   - all buckets [0, max) are inserted into values
+     */
+    internal fun toJsonPayloadObject(): JSONObject {
+        // Include all buckets [0, max), where max is the maximum bucket with
+        // any value recorded.
+        val contiguousValues = if (!values.isEmpty()) {
+            val bucketMax = values.keys.max()!!
+            val contiguousValues = mutableMapOf<String, Long>()
+            for (bucketMin in buckets) {
+                contiguousValues["$bucketMin"] = values.getOrDefault(bucketMin, 0)
```

Comment:
The tests are failing because this requires API 24. Could `values[bucketMin] ?: 0` work in this case?